### PR TITLE
Merge the Tap Service to the Status server

### PR DESCRIPTION
**Pull request by ingwonsong**
Fixes: https://github.com/istio/istio/issues/35833 (to avoid breaking change, 15004 port is still supported in this PR)

The changes are as follows:
- `tap service` (previously only using 15004 port) is served on 15004 as well as the status port (15020).
- `tap_proxy.go` is moved to `pilot/cmd/pilot-agent/status/debugtap/proxy.go`.
- Due to the import cycle, `pilot/cmd/pilot-agent/status/server.go` is moved to `pilot/cmd/pilot-agent/status/server/server.go`, and only the parts which are shared with the other packages are exposed in `pilot/cmd/pilot-agent/status/status.go`.
- `tap service` can be disabled by `PROXY_XDS_DEBUG_VIA_AGENT` on both of 15004 and 15020
- If `PROXY_XDS_DEBUG_VIA_AGENT_PORT` is set to `0`, the status port is only used for serving `tap service`.

cc: @howardjohn

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
This is a breaking change

**Comment by ingwonsong** [comment]
The port number 15004 is not used currently. But, if we need to keep it, I can support 15004 as well for keeping compatibility.

@howardjohn what do you think?

**Comment by ingwonsong** [comment]
To give some transition periods for "just-in-case", `15004` port also serves the debugging interface.
15004 port can be disabled by setting `PROXY_XDS_DEBUG_VIA_AGENT_PORT=0`.

**Comment by ingwonsong** [comment]
https://github.com/istio/istio/pull/40687 would be another way to resolve the dependency cycle issue instead of refactoring "server.go".

@howardjohn Can I get your opinion?

File: pilot/cmd/pilot-agent/status/server/server.go [hunk 1]
```diff
@@ -355,6 +329,24 @@ func (s *Server) Run(ctx context.Context) {
...
 
+	var handler http.Handler = mux
+	if s.config.DebugTapClient != nil {
+		debugTapProxy := debugtap.NewProxy(s.config.DebugTapClient)
+
+		// Registers HTTP Handlers for Debug Tap
+		debugTapProxy.RegisterHTTPHandler(mux)
+		grpcs := grpc.NewServer(istiogrpc.ServerOptions(istiokeepalive.DefaultOption())...)
+		debugTapProxy.RegisterGRPCHandler(grpcs)
+		mixedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/status/server/server.go:340]: do we need the `isRequestFromLocalhost` stuff?

File: pilot/cmd/pilot-agent/app/cmd.go [hunk 2]
```diff
@@ -172,6 +165,14 @@ func newProxyCommand() *cobra.Command {
 			if err != nil {
 				return err
 			}
+
+			// If a status port was provided, start handling status probes.
+			if proxyConfig.StatusPort > 0 {
+				if err := initStatusServer(ctx, proxy, proxyConfig, agentOptions.EnvoyPrometheusPort, agent); err != nil {
```
[the remark this answers is not in the document]
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/app/cmd.go]: I see.

I revert the order of initialization as you commented, and instead, changed for the tap server to get the xds proxy in lazy way.

Please take a look into that once again. Thanks!

File: pilot/cmd/pilot-agent/status/server/server.go (hunk 1 above)
Comment (ingwonsong) [reply pilot/cmd/pilot-agent/status/server/server.go:344]: Good point. Done.

**Pull request closed without merging**